# Host-Origin-and-Parallel-Evolution
Repository for "Host origin is a determinant of parallel evolution between influenza virus gene segments" by Jones &amp; Lakdawala (bioRxiv 2022). Includes raw data and source code for analysis of parallel evolution between gene segments of human and avian influenza viruses.
## How to use this repository:
The code provided in this repository can be used to reconstruct and analyze convergence between phylogenetic trees of gene and protein sequences. Input FASTA files from H9 viruses are provided as examples. Source code for tree reconstruction and analysis is broadly generalizable to any set of alignments and is presented separately from sequence processing and selection. 
# Folders
## 'Data'
This folder contains the raw data that were processed and analyzed in this study. This includes H9 virus FASTA files sourced from the [Influenza Research Database](https://www.fludb.org/). Human H3N2 virus FASTA files were sourced from [an earlier study](https://github.com/Lakdawala-Lab/Parallel-Evolution-Of-Influenza-Viral-RNA/tree/main/Preprocessed%20FASTA%20Files/H3N2%20Virus%20Sequences).
## 'Sequence Selection'
This folder contains source code for the initial processing of raw FASTA files and selection of sequences from different hosts. This code can be run directly with files provided in the Data folder.
## 'Tree Reconstruction and Analysis'
This folder contains source code for tree reconstruction and tree distance calculation. This code can be run with avian H9 virus alignments generated using FASTA files provided the 'Data' folder and source code included in the 'Sequence Selection' folder. Alternatively, this code can be run directly with alignments generated by the user.

### *Happy tree building!* ðŸ§¬
